pkg/gardenlet/controller/shoot/shoot_control_reconcile.go

```diff
@@ -53,11 +56,27 @@ func (c *Controller) runReconcileShootFlow(o *operation.Operation, operationType
 		}
 	}
 
+	var dnsEnabled = !o.Shoot.DisableDNS
+
 	// TODO: timuthy - Only required for migration and can be removed in a future version
-	if o.Shoot.ExternalClusterDomain != nil && garden.DomainIsDefaultDomain(*o.Shoot.ExternalClusterDomain, o.Garden.DefaultDomains) != nil {
-		o.Logger.Info("Migration step - setting primary DNS provider")
-		if err := kutil.SubmitEmptyPatch(context.TODO(), o.K8sGardenClient.Client(), o.Shoot.Info); err != nil {
-			return gardencorev1beta1helper.NewWrappedLastErrors(gardencorev1beta1helper.FormatLastErrDescription(err), err)
+	if dnsEnabled {
+		var (
+			primaryDNSProvider = gardencorev1beta1helper.FindPrimaryDNSProvider(o.Shoot.Info.Spec.DNS.Providers)
+			usesDefaultDomain  = o.Shoot.ExternalClusterDomain != nil && garden.DomainIsDefaultDomain(*o.Shoot.ExternalClusterDomain, o.Garden.DefaultDomains) != nil
```

Comment:
`garden.DomainIsDefaultDomain` should be improved to test for `"."+defaultDomain.Domain` suffix like you already do in the dns admission plugin.
https://github.com/gardener/gardener/blob/90fdb8f3afdf1999b9e3938910cda87d02b6c2f2/pkg/operation/garden/garden.go#L113


Please also add an `Entry` to the corresponding unit test for the problematic case like
```go
Entry("no default domain but with same suffix", "foo.foobar.com", []*garden.Domain{defaultDomain}, BeNil()),
```
here: https://github.com/gardener/gardener/blob/5dfca6bfa03e4ee8486562d38f10f861350d822f/pkg/operation/garden/garden_test.go#L171